File: installing/commands/src/migrateResolutions.ts
import path from 'node:path'

import { confirm } from '@inquirer/prompts'
import { convertResolutionsToOverrides } from '@pnpm/config.parse-overrides'
import { globalInfo, logger } from '@pnpm/logger'
import type { ProjectManifest } from '@pnpm/types'
import { writeProjectManifest } from '@pnpm/workspace.project-manifest-writer'
import { updateWorkspaceManifest } from '@pnpm/workspace.workspace-manifest-writer'
import { isCI } from 'ci-info'

export interface MigrateResolutionsOptions {
  rootProjectManifest: ProjectManifest
  rootProjectManifestDir: string
  workspaceDir: string | undefined
  overrides: Record<string, string> | undefined
  save: boolean | undefined
}

export async function maybeMigrateResolutions (opts: MigrateResolutionsOptions): Promise<Record<string, string> | undefined> {
  if (opts.save === false) return undefined
  if (!opts.rootProjectManifest.resolutions) return undefined

  const hasExistingOverrides = opts.overrides != null && Object.keys(opts.overrides).length > 0
  if (hasExistingOverrides) return undefined

  const canPrompt = !isCI && Boolean(process.stdin.isTTY)

  const { overrides: converted, skipped } = convertResolutionsToOverrides(
    opts.rootProjectManifest.resolutions
  )

  if (Object.keys(converted).length === 0 && skipped.length === 0) return undefined

  if (skipped.length > 0) {
    for (const { selector, reason } of skipped) {
      logger.warn({
        message: `Cannot auto-migrate resolution "${selector}": ${reason}`,
        prefix: opts.rootProjectManifestDir,
      })
    }
  }

  if (Object.keys(converted).length === 0) return undefined

  if (!canPrompt) {
    globalInfo(
      `${Object.keys(converted).length} resolution(s) in package.json could be migrated to overrides in pnpm-workspace.yaml. Re-run in an interactive terminal to proceed.`
    )
    return undefined
  }

  const answer = await confirm({
    message: `Migrate ${Object.keys(converted).length} resolution(s) from package.json to overrides in pnpm-workspace.yaml?`,
    default: true,
  })

  if (!answer) return undefined

  const workspaceDir = opts.workspaceDir ?? opts.rootProjectManifestDir
  await updateWorkspaceManifest(workspaceDir, {
    updatedOverrides: converted,
  })

  const { resolutions: _resolutions, ...manifestWithoutResolutions } = opts.rootProjectManifest
  const manifestPath = path.join(opts.rootProjectManifestDir, 'package.json')
  await writeProjectManifest(manifestPath, manifestWithoutResolutions)

  globalInfo(`Migrated ${Object.keys(converted).length} resolution(s) to overrides in pnpm-workspace.yaml`)

  return converted
}

File: pacquet/crates/cli/src/migrate_resolutions.rs
use std::io::{self, BufRead, IsTerminal, Write};
use std::path::Path;

use indexmap::IndexMap;
use pacquet_config_parse_overrides::{ConversionResult, convert_resolutions_to_overrides};
use pacquet_package_manifest::PackageManifest;

pub struct MigrationOutcome {
    pub converted: IndexMap<String, String>,
}

pub fn maybe_migrate_resolutions(
    manifest: &PackageManifest,
    config_overrides: Option<&IndexMap<String, String>>,
    workspace_yaml_path: &Path,
) -> Option<MigrationOutcome> {
    let value = manifest.value();
    let resolutions = value.get("resolutions")?.as_object()?;
    if resolutions.is_empty() {
        return None;
    }
    if let Some(existing) = config_overrides
        && !existing.is_empty()
    {
        return None;
    }

    let entries: Vec<(String, String)> = resolutions
        .iter()
        .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
        .collect();

    let ConversionResult { overrides, skipped } = convert_resolutions_to_overrides(&entries);

    for skip in &skipped {
        eprintln!("warn: Cannot auto-migrate resolution \"{}\": {}", skip.selector, skip.reason);
    }

    if overrides.is_empty() {
        return None;
    }

    let is_ci = std::env::var("CI").is_ok();
    if is_ci || !io::stdin().is_terminal() {
        eprintln!(
            "info: {} resolution(s) in package.json could be migrated to overrides \
             in pnpm-workspace.yaml. Re-run in an interactive terminal to proceed.",
            overrides.len(),
        );
        return None;
    }

    eprint!(
        "Migrate {} resolution(s) from package.json to overrides in pnpm-workspace.yaml? [Y/n] ",
        overrides.len(),
    );
    let _ = io::stdout().flush();

    let mut answer = String::new();
    if io::stdin().lock().read_line(&mut answer).is_err() {
        return None;
    }
    let trimmed = answer.trim().to_lowercase();
    if !trimmed.is_empty() && trimmed != "y" && trimmed != "yes" {
        return None;
    }

    let converted: IndexMap<String, String> = overrides.into_iter().collect();

    if let Err(err) = write_overrides_to_yaml(workspace_yaml_path, &converted) {
        eprintln!("warn: Failed to write overrides to pnpm-workspace.yaml: {err}");
        return None;
    }

    if let Err(err) = remove_resolutions_from_manifest(manifest) {
        eprintln!("warn: Failed to remove resolutions from package.json: {err}");
        return None;
    }

    eprintln!(
        "info: Migrated {} resolution(s) to overrides in pnpm-workspace.yaml",
        converted.len(),
    );

    Some(MigrationOutcome { converted })
}

fn write_overrides_to_yaml(
    yaml_path: &Path,
    overrides: &IndexMap<String, String>,
) -> Result<(), Box<dyn std::error::Error>> {
    let existing_text = std::fs::read_to_string(yaml_path).unwrap_or_default();
    let mut lines: Vec<String> = existing_text.lines().map(str::to_string).collect();

    let indent = find_indent(&lines);
    lines.push(String::new());
    lines.push("overrides:".to_string());
    for (k, v) in overrides {
        lines.push(format!("{}{}: {}", indent, k, yaml_quote_value(v)));
    }

    let content = lines.join("\n") + "\n";
    std::fs::write(yaml_path, content)?;
    Ok(())
}

fn yaml_quote_value(value: &str) -> String {
    let needs_quoting = value.contains('\'')
        || value.contains('"')
        || value.contains(':')
        || value.contains('#')
        || value.contains('{')
        || value.contains('}')
        || value.contains('[')
        || value.contains(']')
        || value.contains(',')
        || value.contains('&')
        || value.contains('*')
        || value.contains('!')
        || value.contains('|')
        || value.contains('>')
        || value.contains('%')
        || value.contains('@')
        || value.contains('`')
        || value.starts_with(' ')
        || value.ends_with(' ')
        || value.starts_with('-')
        || value.starts_with('?');
    if needs_quoting {
        let escaped = value.replace('\'', "''");
        format!("'{}'", escaped)
    } else {
        value.to_string()
    }
}

fn find_indent(lines: &[String]) -> String {
    for line in lines {
        let trimmed = line.trim_start();
        if trimmed.starts_with('-') || trimmed.contains(':') {
            let leading = line.len() - trimmed.len();
            if leading > 0 {
                return line[..leading].to_string();
            }
        }
    }
    "  ".to_string()
}

fn remove_resolutions_from_manifest(
    manifest: &PackageManifest,
) -> Result<(), Box<dyn std::error::Error>> {
    let mut cloned = manifest.clone();
    if let Some(obj) = cloned.value_mut().as_object_mut() {
        obj.remove("resolutions");
    }
    cloned.save()?;
    Ok(())
}
